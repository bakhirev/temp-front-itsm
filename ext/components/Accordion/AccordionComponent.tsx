import styled from 'styled-components';

export const AccordionComponent = styled.ul`
  list-style-type: none;
  margin: 0;
  padding: 0;
  width: 100%;
`;
