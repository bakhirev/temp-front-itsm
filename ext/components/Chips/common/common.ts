export type Size = 'big' | 'small';
