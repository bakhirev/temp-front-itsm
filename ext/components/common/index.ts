export { DEFAULT_THEME } from './default-theme';
export { KEY_CODES } from './key-codes';
export { Z_INDEXES } from './z-indexes';
export { markerStyle } from './marker';
export { useThemeContext } from './theme-context';
