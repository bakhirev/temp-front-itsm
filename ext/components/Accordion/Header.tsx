import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../common';

export const PADDING_HORIZONTAL = 16;

const HEIGHT = 48;
const HEIGHT_SMALL = 40;

export const Header = styled.div<{ small: boolean }>`
  ${({ small, theme }) => css`
    align-items: center;
    cursor: pointer;
    display: flex;
    height: ${small ? HEIGHT_SMALL : HEIGHT}px;
    justify-content: space-between;
    padding: 0 ${PADDING_HORIZONTAL}px;

    & svg {
      fill: ${theme.color.neutral[50]};
    }

    &:focus,
    &:hover {
      background-color: ${theme.color.opacity.blackHover};
      outline: none;
    }
  `};
`;

Header.defaultProps = {
  theme: DEFAULT_THEME,
};
