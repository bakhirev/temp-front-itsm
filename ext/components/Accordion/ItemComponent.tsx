import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../common';

export const ItemComponent = styled.li`
  ${({ theme }) => css`
    border-top: 1px solid ${theme.color.neutral[20]};
    color: ${theme.color.neutral[90]};
    display: flex;
    flex-direction: column;
    position: relative;

    &:last-child {
      border-bottom: 1px solid ${theme.color.neutral[20]};
    }
  `};
`;

ItemComponent.defaultProps = {
  theme: DEFAULT_THEME,
};
