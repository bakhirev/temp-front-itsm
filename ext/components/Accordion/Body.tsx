import styled from 'styled-components';

import { PADDING_HORIZONTAL } from './Header';

const PADDING_TOP = 10;
const PADDING_TOP_SMALL = 6;

const PADDING_BOTTOM = 14;
const PADDING_BOTTOM_SMALL = 10;

export const Body = styled.div<{ small: boolean }>`
  padding: ${({ small }) => `
    ${small ? PADDING_TOP_SMALL : PADDING_TOP}px
    ${PADDING_HORIZONTAL}px
    ${small ? PADDING_BOTTOM_SMALL : PADDING_BOTTOM}px
  `};
`;
