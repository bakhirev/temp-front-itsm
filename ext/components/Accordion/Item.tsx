import React from 'react';
import type { FC, KeyboardEvent, MouseEvent } from 'react';

import { ReactComponent as ChevronDownOutline } from '../Icons/system/ChevronDownOutline.svg';
import { ReactComponent as ChevronUpOutline } from '../Icons/system/ChevronUpOutline.svg';
import { Subtitle2, Subtitle3 } from '../Typography';
import { KEY_CODES } from '../common';

import { Body } from './Body';
import { Header } from './Header';
import { ItemComponent } from './ItemComponent';

import type { IAccordionListItem } from './index';

const ICON_SIZE = 20;

interface IItemProps extends IAccordionListItem {
  small: boolean;
  onChange: (id: number | string, expanded: boolean) => void;
}

export const Item: FC<IItemProps> = ({ content, expanded = false, id, small, title, onChange }) => {
  const handleFocusPrevent = (event: MouseEvent<HTMLDivElement>) => event.preventDefault();

  const handleKeyDown = (event: KeyboardEvent<HTMLDivElement>) => {
    if (event.keyCode === KEY_CODES.ENTER || event.keyCode === KEY_CODES.SPACEBAR) {
      event.preventDefault();
      onChange(id, !expanded);
    }
  };

  const Title = small ? Subtitle3 : Subtitle2;
  const Icon = expanded ? ChevronUpOutline : ChevronDownOutline;

  return (
    <ItemComponent>
      <Header
        small={small}
        tabIndex={1}
        onClick={() => onChange(id, !expanded)}
        onKeyDown={handleKeyDown}
        onMouseDown={handleFocusPrevent}
      >
        <Title>{title}</Title>
        <Icon height={ICON_SIZE} width={ICON_SIZE} />
      </Header>
      {expanded && <Body small={small}>{content}</Body>}
    </ItemComponent>
  );
};
