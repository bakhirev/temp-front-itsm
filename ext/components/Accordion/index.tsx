import React from 'react';
import type { FC, ReactNode } from 'react';

import { AccordionComponent } from './AccordionComponent';
import { Item } from './Item';

export interface IAccordionProps {
  /** Имя класса для переопределения стилей */
  className?: string;
  /** data-test-id атрибут для тестирования компонента */
  dataTestId?: string;
  /** Список элементов для отображения в аккордеоне */
  list: Array<IAccordionListItem>;
  /** Маленький размер аккордеона */
  small?: boolean;
  /** Коллбэк на изменение состояния элементов аккордеона */
  onChange: (id: number | string, expanded: boolean) => void;
}

export interface IAccordionListItem {
  /** Уникальные идентификатор элемента аккордеона */
  id: number | string;
  /** Заголовок элемента аккордеона */
  title: string;
  /** Контент элемента аккордеона */
  content: ReactNode;
  /** Cостояние элемента аккордеона (открыт/закрыт) */
  expanded?: boolean;
}

export const Accordion: FC<IAccordionProps> = ({
  className,
  dataTestId,
  list,
  small = false,
  onChange,
}) => (
  <AccordionComponent className={className} data-test-id={dataTestId}>
    {list.map((item) => (
      <Item key={item.id} small={small} onChange={onChange} {...item} />
    ))}
  </AccordionComponent>
);
