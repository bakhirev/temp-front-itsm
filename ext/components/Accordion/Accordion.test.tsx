import React from 'react';
import { shallow } from 'enzyme';

import { Accordion } from '../Accordion';
import { KEY_CODES } from '../common';
import type { IAccordionProps } from '../Accordion';

jest.mock('../common', () => ({
  DEFAULT_THEME: {},
  KEY_CODES: {
    ENTER: 13,
    SPACEBAR: 32,
  },
}));

describe('Accordion', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const requiredProps: IAccordionProps = {
    list: [
      {
        id: 0,
        title: 'Title',
        content: 'Content',
      },
    ],
    onChange: jest.fn(),
  };

  it('should render component with required properties', () => {
    const wrapper = shallow(<Accordion {...requiredProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render small component with required properties', () => {
    const wrapper = shallow(<Accordion small {...requiredProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render all items from list', () => {
    const item = requiredProps.list[0];
    const itemsList = new Array(4).fill(item, 0, 4);

    const wrapper = shallow(<Accordion {...requiredProps} list={itemsList} />);
    expect(wrapper.children()).toHaveLength(itemsList.length);
  });

  it("should render item expanded if it is specified in item's configuration", () => {
    const item = requiredProps.list[0];
    const expandedItem = {
      ...item,
      expanded: true,
    };

    const wrapper = shallow(<Accordion {...requiredProps} list={[item, expandedItem]} />);

    const renderedItem = wrapper.childAt(0).dive();
    const renderedOpenedItem = wrapper.childAt(1).dive();

    expect(renderedItem.children()).toHaveLength(1);
    expect(renderedOpenedItem.children()).toHaveLength(2);
  });

  it("should call onChange with item's id and new expanded if user clicks on it", () => {
    const wrapper = shallow(<Accordion {...requiredProps} />);

    const renderedItem = wrapper.childAt(0).dive();
    expect(renderedItem.children()).toHaveLength(1);

    const header = renderedItem.childAt(0);
    header.simulate('click');

    const { list, onChange } = requiredProps;
    expect(onChange).toHaveBeenCalledWith(list[0].id, !list[0].expanded);
  });

  it("should call onChange with item's id and new expanded if user presses enter or space keys on focused item", () => {
    [KEY_CODES.ENTER, KEY_CODES.SPACEBAR].forEach((keyCode) => {
      const wrapper = shallow(<Accordion {...requiredProps} />);

      const renderedItem = wrapper.childAt(0).dive();
      expect(renderedItem.children()).toHaveLength(1);

      const header = renderedItem.childAt(0);
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      header.simulate('keydown', { keyCode, preventDefault: () => {} });

      const { list, onChange } = requiredProps;
      expect(onChange).toHaveBeenCalledWith(list[0].id, !list[0].expanded);
    });
  });
});
