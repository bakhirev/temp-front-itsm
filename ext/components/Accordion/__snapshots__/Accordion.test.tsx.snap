// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should render component with required properties 1`] = `
<styled.ul>
  <Component
    content="Content"
    id={0}
    key="0"
    onChange={[MockFunction]}
    small={false}
    title="Title"
  />
</styled.ul>
`;

exports[`Accordion should render small component with required properties 1`] = `
<styled.ul>
  <Component
    content="Content"
    id={0}
    key="0"
    onChange={[MockFunction]}
    small={true}
    title="Title"
  />
</styled.ul>
`;
